src/instructlab/configuration.py

```diff
@@ -318,12 +313,10 @@ class _generate(BaseModel):
     # model configuration
     model_config = ConfigDict(extra="ignore")
 
-    # required fields
-    model: StrictStr
-    taxonomy_path: StrictStr
-    taxonomy_base: StrictStr
+    model: StrictStr = Field(default_factory=lambda: DEFAULTS.DEFAULT_MODEL)
```

Comment:
I may be wrong but I think that we are doing lambdas whenever the value can change over the life of the process (e.g. in tests where the base path is switched for each test case.) Here is the implementation of `DEFAULT_MODEL`:

```
    @property
    def MODELS_DIR(self) -> str:
        return path.join(self._cache_home, STORAGE_DIR_NAMES.MODELS)

    @property
    def DEFAULT_MODEL(self) -> str:
        return path.join(self.MODELS_DIR, self.GGUF_MODEL_NAME)
```

(So it's not a constant value.)

Reply:
@booxter That's what the `DEFAULTS._reset` method is for. For reference, see how we're using it in our temporary path fixture here:

https://github.com/instructlab/instructlab/blob/426a1941a6fb3edaee3963dcf6c6849bfdaa0165/tests/conftest.py#L79

Reply:
When I replace it with direct reference to the value, I get failures in `test_config.py`:

```
tests/test_backends.py ...............                                                [  7%]
tests/test_config.py FFF........F......                                               [ 16%]
tests/test_lab.py ................................................................... [ 50%]
....                                                                                  [ 52%]
tests/test_lab_config.py ...                                                          [ 53%]
tests/test_lab_diff.py ...............                                                [ 61%]
tests/test_lab_download.py ..                                                         [ 62%]
tests/test_lab_evaluate.py .............                                              [ 68%]
tests/test_lab_init.py ......                                                         [ 71%]
tests/test_lab_model_test.py .                                                        [ 72%]
tests/test_lab_serve.py .......                                                       [ 75%]
tests/test_lab_train.py ............                                                  [ 81%]
tests/test_model_chat.py ...                                                          [ 83%]
tests/test_notebooks.py .                                                             [ 83%]
tests/test_package.py .........s..                                                    [ 89%]
tests/test_utils.py .....................                                             [100%]
...
______________________________ TestConfig.test_default_config _______________________________

self = <tests.test_config.TestConfig object at 0x7f61e5d4f050>
cli_runner = <click.testing.CliRunner object at 0x7f619c924350>

    def test_default_config(self, cli_runner):  # pylint: disable=unused-argument
        cfg = config.Config()
        assert cfg is not None
>       self._assert_defaults(cfg)

/home/vagrant/git/instructlab/tests/test_config.py:109:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.test_config.TestConfig object at 0x7f61e5d4f050>
cfg = Config(chat=_chat(model='/var/tmp/pytest-of-vagrant/pytest-48/test_default_config0/.cache/instructlab/models/merlinite...chbranch(taxonomy_path='/var/tmp/pytest-of-vagrant/pytest-48/test_default_config0/.local/share/instructlab/taxonomy')))

    def _assert_defaults(self, cfg: config.Config):
        # redefine defaults here instead of relyin on those in configuration.DEFAULTS
        # to catch any errors if we are doing things incorrectly over there
        package_name = "instructlab"
        internal_dirname = "internal"
        data_dir = platformdirs.user_data_dir(package_name)
        cache_dir = platformdirs.user_cache_dir(package_name)
        default_model = f"{cache_dir}/models/merlinite-7b-lab-Q4_K_M.gguf"

        assert cfg.general is not None
        assert cfg.version is not None
        assert cfg.general.log_level == "INFO"
        assert cfg.general.debug_level == 0

        assert cfg.chat is not None
        assert cfg.chat.model == default_model
        assert not cfg.chat.vi_mode
        assert cfg.chat.visible_overflow
        assert cfg.chat.context == "default"
        assert cfg.chat.session is None
        assert cfg.chat.logs_dir == f"{data_dir}/chatlogs"
        assert not cfg.chat.greedy_mode

        assert cfg.evaluate is not None

        assert cfg.general is not None
        assert cfg.general.log_level == "INFO"

        assert cfg.generate is not None
        assert cfg.generate.teacher.model_path == default_model
        assert cfg.generate.teacher.llama_cpp is not None
        assert cfg.generate.teacher.llama_cpp.gpu_layers == -1
        assert cfg.generate.teacher.llama_cpp.max_ctx_size == 4096
        assert cfg.generate.teacher.llama_cpp.llm_family == ""
        assert cfg.generate.teacher.vllm is not None
        assert cfg.generate.teacher.vllm.vllm_args == []
        assert cfg.generate.teacher.host_port == "127.0.0.1:8000"
        assert cfg.generate.teacher.backend is None
        assert cfg.generate.teacher.chat_template is None
>       assert cfg.generate.model == default_model
E       AssertionError: assert '/home/vagrant/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf' == '/var/tmp/pytest-of-vagrant/pytest-48/test_default_config0/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf'
E
E         - /var/tmp/pytest-of-vagrant/pytest-48/test_default_config0/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf
E         + /home/vagrant/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf

/home/vagrant/git/instructlab/tests/test_config.py:62: AssertionError
_______________________________ TestConfig.test_cfg_auto_fill _______________________________

self = <tests.test_config.TestConfig object at 0x7f61e5d4ff50>
tmp_path_home = PosixPath('/var/tmp/pytest-of-vagrant/pytest-48/test_cfg_auto_fill0')

        def test_cfg_auto_fill(self, tmp_path_home):
            config_path = tmp_path_home / "config.yaml"
            with open(config_path, "w", encoding="utf-8") as config_file:
                config_file.write(
                    """general:
      log_level: INFO
    """
                )
            cfg = config.read_config(config_path)
>           self._assert_defaults(cfg)

tests/test_config.py:121:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <tests.test_config.TestConfig object at 0x7f61e5d4ff50>
cfg = Config(chat=_chat(model='/var/tmp/pytest-of-vagrant/pytest-48/test_cfg_auto_fill0/.cache/instructlab/models/merlinite-...nchbranch(taxonomy_path='/var/tmp/pytest-of-vagrant/pytest-48/test_cfg_auto_fill0/.local/share/instructlab/taxonomy')))

    def _assert_defaults(self, cfg: config.Config):
        # redefine defaults here instead of relyin on those in configuration.DEFAULTS
        # to catch any errors if we are doing things incorrectly over there
        package_name = "instructlab"
        internal_dirname = "internal"
        data_dir = platformdirs.user_data_dir(package_name)
        cache_dir = platformdirs.user_cache_dir(package_name)
        default_model = f"{cache_dir}/models/merlinite-7b-lab-Q4_K_M.gguf"

        assert cfg.general is not None
        assert cfg.version is not None
        assert cfg.general.log_level == "INFO"
        assert cfg.general.debug_level == 0

        assert cfg.chat is not None
        assert cfg.chat.model == default_model
        assert not cfg.chat.vi_mode
        assert cfg.chat.visible_overflow
        assert cfg.chat.context == "default"
        assert cfg.chat.session is None
        assert cfg.chat.logs_dir == f"{data_dir}/chatlogs"
        assert not cfg.chat.greedy_mode

        assert cfg.evaluate is not None

        assert cfg.general is not None
        assert cfg.general.log_level == "INFO"

        assert cfg.generate is not None
        assert cfg.generate.teacher.model_path == default_model
        assert cfg.generate.teacher.llama_cpp is not None
        assert cfg.generate.teacher.llama_cpp.gpu_layers == -1
        assert cfg.generate.teacher.llama_cpp.max_ctx_size == 4096
        assert cfg.generate.teacher.llama_cpp.llm_family == ""
        assert cfg.generate.teacher.vllm is not None
        assert cfg.generate.teacher.vllm.vllm_args == []
        assert cfg.generate.teacher.host_port == "127.0.0.1:8000"
        assert cfg.generate.teacher.backend is None
        assert cfg.generate.teacher.chat_template is None
>       assert cfg.generate.model == default_model
E       AssertionError: assert '/home/vagrant/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf' == '/var/tmp/pytest-of-vagrant/pytest-48/test_cfg_auto_fill0/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf'
E
E         - /var/tmp/pytest-of-vagrant/pytest-48/test_cfg_auto_fill0/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf
E         + /home/vagrant/.cache/instructlab/models/merlinite-7b-lab-Q4_K_M.gguf

tests/test_config.py:62: AssertionError
```

Reply:
The reason why the failure happens is because class attributes, including their defaults, are evaluated on class definition (which happens when python interpreter imports the module and hits the `class _<model>(...):` section).

Pydantic `Fields` with `default_factory` [allow](https://docs.pydantic.dev/latest/concepts/fields/#default-values) the evaluation to postpone to when the object of the model class is created, which in test case context happens after platform directories are (temporarily) switched for the duration of the test case.